xds/internal/balancer/ringhash/picker.go

```diff
@@ -0,0 +1,140 @@
+		// Shutdown can happen in a race where the old picker is called. A new
+		// picker should already be sent.
+		return true, balancer.PickResult{}, balancer.ErrNoSubConnAvailable
+	default:
+		// Should never reach this. All the connectivity states are already
+		// handled in the cases.
+		return true, balancer.PickResult{}, status.Errorf(codes.Unavailable, "SubConn has undefined connectivity state: %v", state)
+	}
+}
+
+func (p *picker) Pick(info balancer.PickInfo) (balancer.PickResult, error) {
+	e := p.ring.pick(getRequestHash(info.Ctx))
+	if ok, pr, err := handleRICS(e); ok {
+		return pr, err
+	}
+	// ok was false, the entry is in transient failure.
+	return p.handleTransientFailure(e)
+}
+
+func (p *picker) handleTransientFailure(e *ringEntry) (balancer.PickResult, error) {
+	// Queue a connect on the first picked SubConn.
+	e.sc.connect()
+
+	// Find next entry in the ring, skipping duplicate SubConns.
+	e2 := nextSkippingDuplicates(p.ring, e)
+	if e2 == nil {
+		// There's no next entry available, fail the pick.
+		return balancer.PickResult{}, status.Errorf(codes.Unavailable, "the only SubConn is not Ready")
+	}
+
+	// For the second SubConn, also check Ready/IDLE/Connecting as if it's the
+	// first entry.
+	if ok, pr, err := handleRICS(e2); ok {
+		return pr, err
+	}
+
+	// The second SubConn is also in TransientFailure. Queue a connect on it.
+	e2.sc.connect()
+
+	// If it gets here, this is after the second SubConn, and the second SubConn
+	// was in TransientFailure.
+	//
+	// Loop over all other SubConns:
+	// - If all SubConns so far are all TransientFailure, trigger Connect() on
+	// the TransientFailure SubConns, and keep going.
+	// - If there's one SubConn that's not in TransientFailure, keep checking
+	// the remaining SubConns (in case there's a Ready, which will be returned),
+	// but don't not trigger Connect() on the other SubConns.
+	var firstNonFailedFound bool
+	for ee := nextSkippingDuplicates(p.ring, e2); ee != e; ee = nextSkippingDuplicates(p.ring, ee) {
+		scState := ee.sc.effectiveState()
+		if scState == connectivity.Ready {
+			return balancer.PickResult{SubConn: ee.sc.sc}, nil
+		}
+		if firstNonFailedFound {
+			continue
+		}
+		if scState == connectivity.TransientFailure {
+			// This will queue a connect.
+			ee.sc.connect()
+			continue
+		}
+		// This is a SubConn in a non-failure state. We continue to check the
+		// other SubConns, but remember that there was a non-failed SubConn
+		// seen. After this, Pick() will never trigger any SubConn to Connect().
+		firstNonFailedFound = true
+		if scState == connectivity.Idle {
+			// This is the first non-failed SubConn, and it is in a real IDLE
+			// state. Trigger it to Connect().
+			ee.sc.connect()
+		}
+	}
+	return balancer.PickResult{}, status.Errorf(codes.Unavailable, "no connection is Ready")
+}
+
+func nextSkippingDuplicates(ring *ring, cur *ringEntry) *ringEntry {
+	for next := ring.next(cur); next != cur; next = ring.next(next) {
```

Comment:
Should this be `next != ring` ?  Otherwise it seems like this would go forever -- is there a test that exhausts all the subconns and has more than 2?

Reply:
`ring` is not the same type of `next`. This is not a linked list.

`TestNextSkippingDuplicatesOnlyDup` should cover that.

Reply:
Ohh, `cur` is never updated.  I think I saw `next := ring.next(cur)` and thought that was the iterating condition.  Maybe rename `cur` to `entry` since `cur` sounds like something that will change?

Reply:
Done